InternetStoreCHIReloaded/BLL/Services/CategoryService.cs

```diff
@@ -0,0 +1,193 @@
+using BLL.Contracts;
+using BLL.Interfaces;
+using DAL.Entities;
+using DAL.Interfaces;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace BLL.Services
+{
+    public class CategoryService : ICategoryService
+    {
+        private readonly IGenericRepository<CategoryEntity> _categoryRepository;
+        private readonly IMapper _mapper;
+
+        public CategoryService(
+            IGenericRepository<CategoryEntity> categoryRepository,
+            IMapper mapper)
+        {
+            _categoryRepository = categoryRepository;
+            _mapper = mapper;
+        }
+
+        private bool IsValid(Category category)// only name and description
+        {
+            bool validated = true;
+
+            Regex reg;
+
+            reg = new Regex("^[^ ][a-zA-Z ]{3,20}");
+            validated = validated && reg.IsMatch(category.CategoryName);
+
+            reg = new Regex("^[^ ][a-zA-Z0-9. -]{3,200}");
+            validated = validated && reg.IsMatch(category.CategoryDescription);
+
+            return validated;
+        }
+
+        private bool IsPresentInDb(string categoryName)
+        {
+            return _categoryRepository.FindFirstOrDefault(c => c.Name == categoryName) != null;
+        }
+
+        public async Task<ResultContract> AddCategoryAsync(Category newCategory)
+        {
+            var result = new ResultContract();
+
+            if (IsValid(newCategory))
+            {
+                if (!IsPresentInDb(newCategory.CategoryName))
+                {
+                    var category = _mapper.Map<CategoryEntity>(newCategory);
+                    category.CreatedDate = category.UpdatedDate = DateTime.UtcNow;
+
+                    bool success = _categoryRepository.Add(category);
+
+                    if (success)
+                        result.Message = "Category added successfully!";
+                    else
+                        result.Message = "Unexpected error occured during adding new category.";
+
+                    result.IsSuccessful = success;
+                }
+                else
+                {
+                    result.IsSuccessful = false;
+                    result.Message = "Category name is already in database.";
+                }
+            }
+            else
+            {
+                result.IsSuccessful = false;
+                result.Message = "Inappropriate category name or description";
+            }
+
+            return result;
+        }
+
+        public async Task<ResultContract> DeleteCategoryAsync(int id)
+        {
+            var result = new ResultContract();
+            bool success;
+
+            try
+            {
+                success = _categoryRepository.Remove(id);
+            }
+            catch(DbUpdateConcurrencyException)// when trying to delete with id not in Db
```

Comment:
Please move this `try-catch` to the repository level. It's not business-logic responsibility